polyfill/lib/ecmascript.mjs

```diff
@@ -3467,171 +3354,15 @@ export function BalanceTimeDuration(norm, largestUnit) {
-      timeZoneRec,
-      'iso8601',
-      days
-    ).epochNs;
-  }
-
-  const endNs = AddInstant(intermediateNs, norm);
-  norm = TimeDuration.fromEpochNsDiff(endNs, startNs);
-  if (norm.isZero()) {
-    return { days: 0, hours: 0, minutes: 0, seconds: 0, milliseconds: 0, microseconds: 0, nanoseconds: 0 };
-  }
-
-  if (IsCalendarUnit(largestUnit) || largestUnit === 'day') {
-    precalculatedPlainDateTime ??= GetPlainDateTimeFor(timeZoneRec, startInstant, 'iso8601');
-    ({ days, norm } = NormalizedTimeDurationToDays(norm, zonedRelativeTo, timeZoneRec, precalculatedPlainDateTime));
-    largestUnit = 'hour';
-  } else {
-    days = 0;
-  }
-
-  const { hours, minutes, seconds, milliseconds, microseconds, nanoseconds } = BalanceTimeDuration(norm, largestUnit);
-  return { days, hours, minutes, seconds, milliseconds, microseconds, nanoseconds };
-}
-
-export function UnbalanceDateDurationRelative(years, months, weeks, days, largestUnit, plainRelativeTo, calendarRec) {
-  // calendarRec must have looked up dateAdd and dateUntil
-  const TemporalDuration = GetIntrinsic('%Temporal.Duration%');
-  const defaultLargestUnit = DefaultTemporalLargestUnit(years, months, weeks, days, 0, 0, 0, 0, 0, 0);
-  const effectiveLargestUnit = LargerOfTwoTemporalUnits(largestUnit, 'day');
-  if (LargerOfTwoTemporalUnits(defaultLargestUnit, effectiveLargestUnit) === effectiveLargestUnit) {
-    // no-op
-    return { years, months, weeks, days };
-  }
-  if (!calendarRec) throw new RangeError(`a starting point is required for ${largestUnit}s balancing`);
-
-  switch (effectiveLargestUnit) {
-    case 'year':
-      throw new Error('assert not reached');
-    case 'month': {
-      // balance years down to months
-      const later = CalendarDateAdd(calendarRec, plainRelativeTo, new TemporalDuration(years));
-      const untilOptions = ObjectCreate(null);
-      untilOptions.largestUnit = 'month';
-      const untilResult = CalendarDateUntil(calendarRec, plainRelativeTo, later, untilOptions);
-      const yearsInMonths = GetSlot(untilResult, MONTHS);
-      return { years: 0, months: months + yearsInMonths, weeks, days };
-    }
-    case 'week': {
-      // balance years and months down to days
-      const later = CalendarDateAdd(calendarRec, plainRelativeTo, new TemporalDuration(years, months));
-      const yearsMonthsInDays = DaysUntil(plainRelativeTo, later);
-      return { years: 0, months: 0, weeks, days: days + yearsMonthsInDays };
-    }
-    default: {
-      // largestUnit is "day", or any time unit
-      // balance years, months, and weeks down to days
-      const later = CalendarDateAdd(calendarRec, plainRelativeTo, new TemporalDuration(years, months, weeks));
-      const yearsMonthsWeeksInDays = DaysUntil(plainRelativeTo, later);
-      return { years: 0, months: 0, weeks: 0, days: days + yearsMonthsWeeksInDays };
-    }
-  }
-}
+export function UnbalanceDateDurationRelative(years, months, weeks, days, plainRelativeTo, calendarRec) {
```

Comment:
Agreed, but I'd also like to put this off until after #2826 lands. If making that adjustment isn't actually going to change any observable results, I'd prefer to keep Duration.compare as it is for now, since it's otherwise not affected by this optimization.

Reply:
→ #2837